src/librustdoc/clean/types.rs

```diff
@@ -1968,9 +1969,55 @@ crate struct Static {
 #[derive(Clone, PartialEq, Eq, Hash, Debug)]
 crate struct Constant {
     crate type_: Type,
-    crate expr: String,
-    crate value: Option<String>,
-    crate is_literal: bool,
+    crate kind: ConstantKind,
+}
+
+#[derive(Clone, PartialEq, Eq, Hash, Debug)]
+crate enum ConstantKind {
+    /// This is the wrapper around `ty::Const` for a non-local constant. Because it doesn't have a
+    /// `BodyId`, we need to handle it on its own.
+    TyConst { expr: String },
+    /// A constant (expression) that's not an item or associated item. These are usually found
+    /// nested inside types (e.g., array lengths) or expressions (e.g., repeat counts), and also
+    /// used to define explicit discriminant values for enum variants.
+    Anonymous { body: BodyId },
+    /// A constant from a different crate.
+    Extern { def_id: DefId },
```

Comment:
I feel confused about this: `TyConst` says (emphasis mine)

> This is the wrapper around `ty::Const` for a **non-local constant**.

but `Extern` *also* says

> A constant from a **different crate**.

What's the difference between a "non-local constant" and a "constant from a different crate"?

Reply:
`TyConst` is about generics (meaning, "non local") whereas `Extern` is for foreign constants items.

Reply:
When you say "about generics" do you mean it's about const generics? What's an example of a `TyConst` and of an `Extern`?

Reply:
You can only find `TyConst` when listing over an item generic arguments. An extern is simply a const variable defined in another crate.

Reply:
Take a look at `fn external_generic_args` in `clean/utils.rs`.

Reply:
Why do we need to track `Extern` separately from `Local`? It seems like you can get the value of the `Extern` just from the `DefId`, so can't we merge `Extern` and `Local`?

Reply:
The way they're displayed is both different (one use a `BodyId` whereas the other is forced to use its `DefId`. But ultimately, it'd be nice to be able to merge them. But I think it's out of the scope of this PR.